build.rs

```diff
@@ -119,16 +160,76 @@ mod generate {
 
             as_enum.write_all(b"}\n")?;
 
-            as_enum.write_all(b"impl DrmFormat {\n")?;
-            as_enum.write_all(b"pub(crate) fn from_u32(n: u32) -> Option<Self> {\n")?;
+            writeln!(as_enum, "impl {} {{", name)?;
+            writeln!(as_enum, "pub(crate) fn from_{}(n: {}) -> Option<Self> {{\n", repr, repr)?;
             as_enum.write_all(b"match n {\n")?;
 
             for (member, value) in &members {
                 writeln!(as_enum, "{} => Some(Self::{}),", value, member)?;
             }
 
             writeln!(as_enum, "_ => None")?;
-            as_enum.write_all(b"}}}")?;
+            as_enum.write_all(b"}}}\n")?;
+
+            Ok(())
+        }
+
+        let as_enum_path = "src/as_enum.rs";
+        {
+            let mut as_enum = File::create(as_enum_path)?;
+
+            as_enum.write_all(b"// Automatically generated by build.rs\n")?;
+            as_enum.write_all(b"use crate::consts;")?;
+
+            write_enum(&mut as_enum, "DrmFormat", "u32", format_names)?;
+
+            as_enum.write_all(b"#[derive(Debug)]")?;
+            write_enum(&mut as_enum, "DrmVendor", "u8", vendor_names)?;
+
+            // modifiers can overlap
+
+            as_enum.write_all(b"#[derive(Debug, Copy, Clone)]")?;
+            as_enum.write_all(
+                b"#[cfg_attr(feature = \"serde\", derive(serde::Serialize, serde::Deserialize))]",
+            )?;
+            as_enum.write_all(b"pub enum DrmModifier {\n")?;
+
+            let modifier_members: Vec<(String, String)> = modifier_names
+                .iter()
+                .map(|(_, short)| {
+                    (
+                        enum_member_case(short),
+                        format!("consts::{}{}", CONST_PREFIX, short),
+                    )
+                })
+                .collect();
+            for (member, _) in &modifier_members {
+                writeln!(as_enum, "{},", member)?;
+            }
+            as_enum.write_all(b"Unrecognized(u64)")?;
+
+            as_enum.write_all(b"}\n")?;
+
+            as_enum.write_all(b"impl DrmModifier {\n")?;
+            as_enum.write_all(b"pub(crate) fn from_u64(n: u64) -> Self {\n")?;
+            as_enum.write_all(b"#[allow(unreachable_patterns)]\n")?;
```

Comment:
I don't see any warning without this. Why is it necessary?

Reply:
Because they can overlap as some vendor-specific modifiers might just be generic-modifiers.

E.g. with my kernel version, I am getting:
```
warning: unreachable pattern
   --> src/as_enum.rs:316:13
    |
316 |             consts::DRM_FOURCC_SAMSUNG_16_16_TILE => Self::Samsung_16_16_tile,
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unreachable_patterns)]` on by default

warning: 1 warning emitted
```
Because `DRM_FOURCC_SAMSUNG_16_16_TILE` has the same value as `DRM_FOURCC_GENERIC_16_16_TILE`.

That means the match statement will contain the same value two times, which makes the second match unreachable. Currently sorting defines, which match arm is first and will always return, but because of the custom `PartialEq` implementation, I figured that it does not really matter.